### ISSUE TITLE
加载模型后运行 model->test(); 返回错误

### ISSUE THREAD

**Issue by heng540440** [espressif/esp-dl#303]
### Checklist

- [x] Checked the issue tracker for similar issues to ensure this is not a duplicate.
- [x] Provided a clear description of your suggestion.
- [x] Included any relevant context or examples.

### Issue or Suggestion Description

加载模型成功运行model->test() 返回错误
I (396) TMF8829_APP: Model loaded successfully!
I (396) TMF8829_APP: Input size: 1024 bytes, type: float
I (396) TMF8829_APP: Output size: 9 bytes, type: float

I (446) dl::Model: Testing output distance.
E (446) compare_elements: Inconsistent values, ground truth: 0.9999939203, infer: 0.9975273609, epsilon:0.0000200000
E (446) compare_elements: The position is: [0, 1]
E (456) dl::Model: Test output distance does not match

E (456) TMF8829_APP: ******************mode test err*****************************

还有一个问题 我的模型输入是一个float型的32*32的数组，输出是float型长度为9的数组
    // 打印基础信息
    ESP_LOGI(TAG, "Input size: %d bytes, type: %s", 
             input_tensor->get_size(), input_tensor->get_dtype_string());
    ESP_LOGI(TAG, "Output size: %d bytes, type: %s",
             output_tensor->get_size(), output_tensor->get_dtype_string());
输出结果是
I (396) TMF8829_APP: Input size: 1024 bytes, type: float
I (396) TMF8829_APP: Output size: 9 bytes, type: float
如果类型是float 那么size应该是4096bytes 这里为什么是1024 bytes包括输出也是 这对吗？

现在模型运行结果不正常
我使用如下方式获取模型输出结果
float *out = (float *)output_tensor->get_element_ptr();

结果打印：
I (6405) TMF8829_APP:   [0]: 0.000000
I (6405) TMF8829_APP:   [1]: 0.997527
I (6405) TMF8829_APP:   [2]: 0.000000
I (6405) TMF8829_APP:   [3]: 0.000000
I (6405) TMF8829_APP:   [4]: 0.000000
I (6405) TMF8829_APP:   [5]: 0.000000
I (6415) TMF8829_APP:   [6]: 0.000000
I (6415) TMF8829_APP:   [7]: 0.002473
I (6415) TMF8829_APP:   [8]: 0.000000

此时状态应该是TMF8829_APP:   [0]:  比较接近1才正确

**Comment by 100312dog** [comment]
@heng540440 可以传一下模型onnx文件不

**Comment by heng540440** [comment]
> [@heng540440](https://github.com/heng540440) 可以传一下模型onnx文件不
你是说量化onnx的python脚本吗？onnx模型文件上传不了呀

这是我的量化脚本
"""
ESP-DL 模型量化脚本
将 ONNX 模型量化后导出为 ESP-DL 可用的 .espdl 文件
"""
from torch.utils.data import DataLoader, TensorDataset
from esp_ppq.api import espdl_quantize_onnx
import numpy as np
import torch
import glob
import os
from esp_ppq import QuantizationSettingFactory, QuantizationSetting
from esp_ppq.api import espdl_quantize_onnx, get_target_platform


def generate_my_data(num_samples=200, data_dir='date', return_single=False):
    """
    从真实 CSV 文件加载校准数据

    Args:
        num_samples: 采样的样本数量，默认200
        data_dir: CSV文件所在目录，默认'date'

    Returns:
        torch.Tensor: shape为[num_samples, 32, 32]的float32张量
    """
    # 查找目录下所有CSV文件
    csv_files = glob.glob(os.path.join(data_dir, '*.csv'))
    if not csv_files:
        raise FileNotFoundError(f"在 {data_dir} 目录下未找到 CSV 文件")

    all_features = []
    for csv_file in csv_files:
        # 加载CSV，跳过第一行（通常是表头）
        data = np.loadtxt(csv_file, delimiter=',', skiprows=1)
        # 处理单行数据情况（只有一行数据）
        if data.ndim == 1:
            data = data.reshape(1, -1)
        # CSV格式：前1024列是特征，最后一列是标签
        features = data[:, :-1]  # 取前1024个特征
        all_features.append(features)

    # 合并所有CSV的特征数据
    X = np.vstack(all_features)
    # 重塑为32x32图像格式 (行优先：左到右，上到下)
    X = X.reshape(-1, 32, 32)
    # 转换为float32，与Keras模型输入类型一致
    X = X.astype(np.float32)

    X = X / 1000.0
    if return_single:
        # 返回第二个样本，shape: [1, 32, 32]
        X = X[1:6, :, :]
        return torch.tensor(X)
    else:
        # 随机打乱数据并返回指定数量
        indices = np.random.permutation(len(X))[:num_samples]
        X = X[indices]
        return torch.tensor(X)


def collate_fn(batch):
    """
    数据加载器的批处理整理函数

    Args:
        batch: DataLoader返回的批次列表，形状为[(x, y), (x, y), ...]
               由于TensorDataset存储了(x, y)，每个元素是元组

    Returns:
        torch.Tensor: 只需要返回x（特征），丢弃y（标签）
    """
    # batch[0] 是第一个样本的 (x, y) 元组
    # batch[0][0] 取出 x（特征）
    # .to(DEVICE) 将数据放到指定设备（cpu/cuda）
    return batch[0][0].to(DEVICE)


if __name__ == "__main__":
    # ==================== 配置参数 ====================
    ONNX_MODEL_PATH = "distance_classifier.onnx"        # ONNX模型路径
    ESPDL_MODEL_PATH = "distance_classifier.espdl"      # 量化后ESP-DL模型输出路径
    INPUT_SHAPE = [1, 32, 32, 1]                        # 输入形状 [批次, 高度, 宽度, 通道]
    TARGET = "c"                                        # 量化目标: 'c' / 'esp32s3' / 'esp32p4'
    NUM_OF_BITS = 8                                     # 量化位数，int8为8
    DEVICE = "cpu"                                     # 计算设备: 'cuda' 或 'cpu'

    # ==================== 生成校准数据 ====================
    # 校准数据用于量化时统计各层数值范围，必须与真实输入分布一致
    x = generate_my_data(num_samples=200, data_dir='date')

    # 创建PyTorch数据集和DataLoader
    # batch_size=1 因为量化时逐样本推理
    dataset = TensorDataset(x)  # 只传入x，不需要y（标签）
    dataloader = DataLoader(dataset, batch_size=1, shuffle=False)


    # ==================== 生成固定的测试数据（用于 model->test()）====================
    # 使用一个非负的真实样本，确保测试数据与实际输入分布一致
    fixed_test_input = generate_my_data(return_single=True, data_dir='date')
    # 添加通道维度: [1, 32, 32] -> [1, 32, 32, 1]
    fixed_test_input = fixed_test_input.unsqueeze(-1)
    print(f"测试数据 shape: {fixed_test_input.shape}")
    print(f"测试数据范围: min={fixed_test_input.min().item():.4f}, max={fixed_test_input.max().item():.4f}")

    #quant_setting = QuantizationSettingFactory.default_setting()
    quant_setting = QuantizationSettingFactory.espdl_setting()


        # 1. 让所有 Conv 和 MatMul 使用 16bit（提高精度）
    quant_setting.dispatching_table.append(
        ".*/Conv$",
        get_target_platform(TARGET, 16)
    )

    quant_setting.dispatching_table.append(
        ".*/MatMul$",
        get_target_platform(TARGET, 16)
    )

    #     # 3. Softmax 保持 FP32（确保输出概率精度）
    # quant_setting.dispatching_table.append(
    #     ".*/Softmax",
    #     get_target_platform(TARGET, 32)
    # )
    
    # # 4. 可选：输出层全精度
    # quant_setting.dispatching_table.append(
    #     "functional_1/distance_1/.*",
    #     get_target_platform(TARGET, 32)
    # )

    # # 让所有 bias 使用 32bit（完全不量化）
    # quant_setting.dispatching_table.append(
    #     ".*/BiasAdd/ReadVariableOp:0",
    #     get_target_platform(TARGET, 32)
    # )

    #     # 匹配 BiasAdd 对应的 Add 算子
    # quant_setting.dispatching_table.append(
    #     ".*/BiasAdd",  # 注意：没有 ReadVariableOp
    #     get_target_platform(TARGET, 32)
    # )
        

    # ==================== 执行量化 ====================
    quant_ppq_graph = espdl_quantize_onnx(
        onnx_import_file=ONNX_MODEL_PATH,       # 输入：ONNX模型路径
        espdl_export_file=ESPDL_MODEL_PATH,     # 输出：量化后ESP-DL模型路径
        calib_dataloader=dataloader,            # 校准数据加载器
        calib_steps=32,                         # 校准步数（使用32批数据统计量化参数）
        input_shape=INPUT_SHAPE,                 # 模型输入形状
        inputs=[fixed_test_input],                            # 指定输入张量，None自动推断
        #inputs=None,
        target=TARGET,                         # 目标硬件平台
        num_of_bits=NUM_OF_BITS,               # 量化位数
        collate_fn=collate_fn,                  # 批处理整理函数（提取特征x）
        dispatching_override=None,              # 算子分发覆盖，None使用默认
        device=DEVICE,                          # 计算设备
        error_report=True,                      # 输出量化错误报告
        skip_export=False,                      # False=执行量化并导出模型
        export_test_values=True,                # 导出时包含测试值（用于验证）
        setting=quant_setting,  # <--- 关键：在这里传入配置对象
        verbose=1,                               # 日志详细程度 1=详细
    )

**Comment by 100312dog** [comment]
@heng540440 需要onnx文件，可以改个扩展名，比如 model.onnx->model.txt  应该就能传了

**Comment by heng540440** [comment]
> [@heng540440](https://github.com/heng540440) 需要onnx文件，可以改个扩展名，比如 model.onnx->model.txt 应该就能传了

[distance_classifier.onnx.txt](https://github.com/user-attachments/files/28244757/distance_classifier.onnx.txt)

**Comment by heng540440** [comment]
我的量化脚本 
当我使用  quant_setting = QuantizationSettingFactory.espdl_setting()时
model->test(); 是OK的 但是此时模型推理结果错误
然后我加上了
    # 1. Conv层用16bit（提高特征提取精度）
    quant_setting.dispatching_table.append(
        ".*/functional_1/conv2d_1/Relu:0",
        get_target_platform(TARGET, 16)
    )
    quant_setting.dispatching_table.append(
        ".*/functional_1/conv2d_1_2/Relu:0",
        get_target_platform(TARGET, 16)
    )
    quant_setting.dispatching_table.append(
        ".*/functional_1/conv2d_2_1/Relu:0",
        get_target_platform(TARGET, 16)
    )
这时model->test(); 错误 模型推理结果也错误

**Comment by BlueSkyB** [comment]
你是在什么芯片上跑的模型？

**Comment by heng540440** [comment]
> 你是在什么芯片上跑的模型？
esp32c3

**Comment by 100312dog** [comment]
@heng540440 

1. 更新下 esp-ppq 更新到 1.3.2。
2. 
https://github.com/espressif/esp-dl/blob/master/esp-dl/dl/model/src/dl_memory_manager_greedy.cpp#L420
改为
```
#if CONFIG_SPIRAM
    tensor->set_offset(chunk->offset);
#else
    tensor->set_internal_offset(chunk->offset);
#endif
```

晚些esp-dl会更新。
